src/libaktualizr-posix/ipuptanesecondary.cc

```diff
@@ -120,34 +134,77 @@ bool IpUptaneSecondary::putMetadata(const RawMetaPack& meta_pack) {
   return r->result == AKInstallationResult_success;
 }
 
-bool IpUptaneSecondary::sendFirmware(const std::string& data) {
-  std::lock_guard<std::mutex> l(install_mutex);
-  LOG_INFO << "Sending firmware to the secondary";
+Manifest IpUptaneSecondary::getManifest() const {
+  LOG_DEBUG << "Getting the manifest from secondary with serial " << getSerial();
   Asn1Message::Ptr req(Asn1Message::Empty());
-  req->present(AKIpUptaneMes_PR_sendFirmwareReq);
 
-  auto m = req->sendFirmwareReq();
-  SetString(&m->firmware, data);
+  req->present(AKIpUptaneMes_PR_manifestReq);
+
   auto resp = Asn1Rpc(req, getAddr());
 
-  if (resp->present() != AKIpUptaneMes_PR_sendFirmwareResp) {
-    LOG_ERROR << "Failed to get response to sending firmware to secondary";
-    return false;
+  if (resp->present() != AKIpUptaneMes_PR_manifestResp) {
+    LOG_ERROR << "Failed to get a response to a get manifest request to secondary";
+    return Json::Value();
   }
+  auto r = resp->manifestResp();
 
-  auto r = resp->sendFirmwareResp();
-  return r->result == AKInstallationResult_success;
+  if (r->manifest.present != manifest_PR_json) {
+    LOG_ERROR << "Manifest wasn't in json format";
+    return Json::Value();
+  }
+  std::string manifest = ToString(r->manifest.choice.json);  // NOLINT
+  return Utils::parseJSON(manifest);
+}
+
+bool IpUptaneSecondary::ping() const {
+  Asn1Message::Ptr req(Asn1Message::Empty());
+  req->present(AKIpUptaneMes_PR_getInfoReq);
+
+  auto m = req->getInfoReq();
+
+  auto resp = Asn1Rpc(req, getAddr());
+
+  return resp->present() == AKIpUptaneMes_PR_getInfoResp;
 }
 
-data::ResultCode::Numeric IpUptaneSecondary::install(const std::string& target_name) {
-  LOG_INFO << "Invoking an installation of the target on the secondary: " << target_name;
+data::ResultCode::Numeric IpUptaneSecondary::install(const Uptane::Target& target) {
+  std::lock_guard<std::mutex> l(install_mutex);
+  auto install_result = install_v2(target);
+  if (install_result == data::ResultCode::Numeric::kUnknown) {
+    // fallback to the previous installation version
+    LOG_ERROR << "Failed to install " << target.filename() << " on Secondary " << getSerial()
+              << "\nTry to install by applying the previous version of the installation procedure";
+
+    install_result = install_v1(target);
```

Comment:
Can we find a better way to determine which installation method to use? I'd prefer if we could check once and then store the result (not in the database, just in the class). Otherwise this feels like a lot of wasted installation attempts. I know that will probably get covered by the MQTT work, so we might be able to punt on this for now.

Reply:
> Can we find a better way to determine which installation method to use?

Yes, as we discussed it, it requires the proper implementation of the discovery mechanism. Once Secondary is "discovered" it could inform Primary about an API version it supports. There are lots of small items here but in general, IMHO, it does not make sense to do it until we switch to the MQTT based-implementation. What probably can be done now is to add addition RPC request that Primary can perform at the secondary initialization phase and it this requests fails then Primary assumes that the given Secondary is the "old" one if it is successful than it gets the Secondary supported version from the response to that additional RPC request.